MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,236 @@
+/// When we prune node 2, the MMR looks as follows
+///
+///                 /\
+///                /  \
+///               /    \
+///              /      \
+///             /        \
+///            /          \
+///           /            \
+///          /\             \
+///         /\ \            /\
+///        /  \ \          /  \
+///       /\   \ \        /\   \
+///      /  \   \ \      /  \   \
+///     /\  /\  /\ \    /\  /\  /\
+///      /\/\/\/\/\/\  /\/\/\/\/\/\
+///
+/// Although we have no removed node 1 and node 2 from the MMR we cannot yet remove node 3 as we require node 3 for the proof of node 6. Lets prune 4 and 5.
+///
+///                 /\
+///                /  \
+///               /    \
+///              /      \
+///             /        \
+///            /          \
+///           /            \
+///          /\             \
+///         /\ \            /\
+///        /  \ \          /  \
+///       /\   \ \        /\   \
+///      /  \   \ \      /  \   \
+///         /\  /\ \    /\  /\  /\
+///        /\/\/\/\/\  /\/\/\/\/\/\
+///
+/// Now we removed 3 from the MMR
+
+
+use crate::merklenode::MerkleNode;
+
+type ObjectHash = [u8; 32];
+
+pub struct MerkleMountainRange<T>
+where T: Hashable
+{
+    mmr: Vec<MerkleNode>,
+    data: HashMap<ObjectHash, T>,
+    blockheight: Vec<u64>,
+
+}
+
+impl MerkleMountainRange {
+    /// This function creates a new empty Merkle Mountain Range
+    pub fn new() -> MerkleMountainRange {
+        mmr: Vec::new();
+        data: HashMap::new();
+    }
+
+    pub fn add_vec(&self, objects : Vec<T>)
+    where T: Hashable {
+        for i in 0..objects.len(){
+            self.add_single(objects[i]);
+        }
+    }
+
+    /// This function adds a new node to the mmr.
+    pub fn add_single(&self,object: T)
+    where T: Hashable {
+        let node : merklenode{
+            object.get_hash(),
+            false,
+        };
+        self.mmr.append(node);
+        self.data.insert(object.get_hash, object);
+        if is_node_right(self.mmr.len()-1){
+            self.add_single_no_leaf(self.mmr.len()-1)
+        }
+    }
+
+    //This function adds non leaf nodes, eg nodes that are not directly a hash of data
+    //This is iterative and will continue to add up and till it hits the top, will be a future left child
+    fn add_single_no_leaf(&self, index : u64)
+    {
+        let new_hash = self.mmr[index].hash.clone() + self.mmr[pow(2,index+1) -1].hash.clone();
+        let new_node = : merklenode{
+            new_hash,
+            false,
+        };
+        if is_node_right(self.mmr.len()-1){
+            self.add_single_no_leaf(self.mmr.len()-1)
+        }
+     }
+
+/// This function takes in the index and calculates if the node is right child node or not.
+/// If the node is the tree root it will still give the answer as if is a child of a node
+    pub fn is_node_right(index : u64) -> boolean
```

Comment:
This function doesn't do what it says? Guessing it is a placeholder but put a TODO comment otherwise it looks like its complete.